feat: get subscribers (#1391)
WalletConnect/walletconnect-docs: docs/web3inbox/backend-integration.mdx

## Get subscribers

You can tell if an account is subscribed and get information about subscribers using the `/v1/<project-id>/subscribers` endpoint. You can provide up to 100 accounts in the request.

If the account is subscribed, it will be returned as a key in the response along with a list of subscribed notification types as the value. If the account is not a subscriber, it will not be returned in the response.

Example usage:

<Tabs queryString={'api-client'}>

<TabItem value="javascript" label="JavaScript">

```typescript
const PROJECT_ID = '<PROJECT_ID>'
const NOTIFY_API_SECRET = '<NOTIFY_API_SECRET>'
const requestBody: RequestBody = {
  accounts: [
    'eip155:1:0x9AfEaC202C837df470b5A145e0EfD6a574B21029',
    'eip155:1:0x0000000000000000000000000000000000000000'
  ]
}

const response = await fetch(`https://notify.walletconnect.com/v1/${PROJECT_ID}/subscribers`, {
  method: 'POST',
  headers: {
    Authorization: `Bearer ${NOTIFY_API_SECRET}`
  }
  body: JSON.stringify(requestBody)

> })

Reviewer: Why is it a `POST` request ? Is it to circumvent the limit if you were using query params ?

## Get subscribers

You can tell if an account is subscribed and get information about subscribers using the `/v1/<project-id>/subscribers` endpoint. You can provide up to 100 accounts in the request.

> If the account is not a subscriber, it will be `null`. If the account is subscribed, it will include a list of notification types the subscriber has opted-into receiving.

Author: ```suggestion
If the account is subscribed, it will be returned as a key in the response along with a list of subscribed notification types as the value. If the account is not a subscriber, it will not be returned in the response.
```

Example usage:

<Tabs queryString={'api-client'}>

<TabItem value="javascript" label="JavaScript">

```typescript
const PROJECT_ID = '<PROJECT_ID>'
const NOTIFY_API_SECRET = '<NOTIFY_API_SECRET>'
const requestBody: RequestBody = {
  accounts: [
    'eip155:1:0x9AfEaC202C837df470b5A145e0EfD6a574B21029',
    'eip155:1:0x0000000000000000000000000000000000000000'
  ]
}

const response = await fetch(`https://notify.walletconnect.com/v1/${PROJECT_ID}/subscribers`, {
  method: 'POST',
  headers: {
    Authorization: `Bearer ${NOTIFY_API_SECRET}`
  }
  body: JSON.stringify(requestBody)
})

const subscribers: ResponseBody = await response.json()

type RequestBody = {
  // Max 100 accounts
  accounts: string[]
}
type ResponseBody = {

>   [account: string]: Subscriber | null

Author: ```suggestion
  [account: string]: Subscriber
```

}
type Subscriber = {
  notification_types: string[]
}
```

</TabItem>

<TabItem value="curl" label="cURL">

```bash
curl -X POST 'https://notify.walletconnect.com/v1/<PROJECT_ID>/subscribers' \
  --header 'Authorization: Bearer <NOTIFY_API_SECRET>' \
  --header 'Content-Type: application/json' \
  --data '{
    "accounts": [
      "eip155:1:0x9AfEaC202C837df470b5A145e0EfD6a574B21029",
      "eip155:1:0x0000000000000000000000000000000000000000"
    ]
  }'
```

</TabItem>

</Tabs>

Example response:

```jsonc
{
  "eip155:1:0x9AfEaC202C837df470b5A145e0EfD6a574B21029": {
    "notification_types": ["4d1c97ad-c182-4097-8f2c-8f80c0674df2"]

>   },
>   "eip155:1:0x0000000000000000000000000000000000000000": null

Author: ```suggestion
  }
```

Rate limits are implemented as [token bucket](https://en.wikipedia.org/wiki/Token_bucket) and contain both rate and burst amounts. On average, a rate of requests can be made. However, since real-world applications often make requests in bursts, this fixed rate can be surpassed temporarily up to the burst amount, provided the app subsequently makes requests below the average in order to recover its bursting capability.

- `POST /<project-id>/notify`:
  - Each app can send 2 notifications per hour to an account, with a burst up to 50. Accounts that have been rate limited will be returned in the request response. Exceptions may be made on a per-project basis for special circumstances.
  - Each app can call this endpoint 2 times per second with a burst up to 20. Rate limited requests will return a 429 status code.
- `POST /v1/<project-id>/get-subscribers`

>   - Each app can call this endpoint 100 times per second with a burst up to 100. Rate limited requests will return a 429 status code.

Author: I'm taking a stab at some numbers for the limits, 100 accounts, 100/s, and 100 burst seem somewhat sensible

- `GET /<project-id>/subscribers`

>   - Each app can call this endpoint 1 time every 5 minutes with a burst up to 2. Rate limited requests will return a 429 status code.

Author: Note: rate limit has been reduced. I am working with developers to switch to new endpoint before enforcing the new rate limit.

Reviewer 2: Can we be explicit about what qualifies as a burst here? I'm a bit confused (and suspect external devs may be) how a burst is counted.

Assuming 1 per 5 minutes but we'll allow 2 if within some short interval?

Author: This question is asked a lot. I linked to the token bucket Wikipedia article above and tried to summarize it, but not doing a good job

Author: Made an issue: https://github.com/WalletConnect/walletconnect-docs/issues/1401
